README
NAME
    Test::Virtual::Filesystem - Validate a filesystem

SYNOPSIS
        use Test::Virtual::Filesystem;
        Test::Virtual::Filesystem->new({mountdir => '/path/to/test'})->runtests;

    or with more customization:

        use Test::Virtual::Filesystem;
        my $test = Test::Virtual::Filesystem->new({mountdir => '/path/to/test', compatible => '0.03'});
        $test->enable_test_xattr(1);
        $test->enable_test_chown(1);
        $test->enable_test_atime(1);
        $test->runtests;

    See the file t/filesys.t in this distribution or the file t/fusepdf.t in
    the Fuse::PDF distribution for thorough examples.

    WARNING: all of the files in the `mountdir' will be deleted in the
    `teardown' method so BE CAREFUL that you specify the right folder!

LICENSE
    Copyright 2007 Chris Dolan, *cdolan@cpan.org*

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

DESCRIPTION
    If you are creating a filesystem, say via Fuse or Filesys::Virtual, you
    need a fairly mundane set of tests to try out lots of typical filesystem
    operations. This package attempts to accumulate a bunch of those tests
    into a handy suite to make it easier for you to test your filesystem.

    This suite is based on `Test::Class', a fantastic library for organizing
    tests into bite-sized bundles. The power of Test::Class lets you select
    a subset of tests to run at author time. For example, when I was working
    on the extended attribute (aka `xattr') tests, I found myself typing
    this:

      env TEST_METHOD='xattr_.*' perl -Ilib t/filesys.t

    which runs just the test methods that begin with `xattr_'.

    There are several methods that let you turn on or off a subset of the
    tests. For example, if you do not intend that your filesystem will
    support symbolic links, you can invoke `$test-'enable_test_symlink(0)>
    in your test program just before you call `$test-'runtests>.

COMPATIBILITY POLICY
    Every time I add a new test to this suite, I annotate it with a version
    number. If client code specifies an expected version number (say, 1.10)
    and it's running against a newer version or this module (say, 1.20) then
    any newer test will be marked as a TODO test. That way if the test
    fails, it won't regress published code that used to work.

    This policy will allow us to continue adding new filesystem tests
    without worrying about breaking existing CPAN modules.

CAVEATS AND LIMITATIONS
    This module needs a more complete suite of test cases. In particular,
    tests are needed for the following filesystem features:

        hardlinks
        nlink
        seek/rewinddir, tell/telldir
        read, sysread, syswrite
        overwrite (with open '+<')
        deep directories
        very full directories
        large files
        non-ASCII filenames (maybe constructor should specify the encoding?)
        permissions
        special file types (fifos, sockets, character and block devices, etc)
        chown
        binmode, non-binmode
        eof
        fileno
        statfs (AKA `df` or `mount`)
        rename corner cases:
         * dest inside src
         * src or dest leaf is '.' or '..'
         * src or dest is FS root
         * dest leaf is symlink
        threading and re-entrancy
        file locking?
        async I/O??

    Any help writing tests (or adapting tests from existing suites) will be
    appreciated!

METHODS
    This module is a subclass of Test::Class. All methods from that class
    are available, particularly `runtests()'.

    $pkg->new({mountdir => $mountdir, ...})
        Create a new test suite which will operate on files contained within
        the specified mount directory. WARNING: any and all files and
        folders in that mount directory will be deleted!

        The supported options are:

        `mountdir'
            This required property indicates where tests should run.

        `compatible'
            Specify a Test::Virtual::Filesystem version number that is known
            to work. If the actual Test::Virtual::Filesystem version number
            is greater, then any test cases added after the specified
            compatible version are considered `TODO' tests. See Test::More
            for details about `TODO' tests.

    $self->init()
        Invoked just before then end of `new()'. This exists solely for
        subclassing convenience. This implementation does nothing.

PROPERTIES
    The following accessor/mutator methods exist to turn on/off various
    features. They all behave in usual Perl fashion: with no argument, they
    return the current value. With one argument, they set the current value
    and return the newly set value.

    $self->enable_test_all()
        As a getter, checks whether all of the other tests are enabled.

        As a setter, turns on/off all the tests.

    $self->enable_test_xattr()
        Default false.

    $self->enable_test_time()
        Default true. If set false, it also sets `atime', `mtime' and
        `ctime' false.

    $self->enable_test_atime()
        Default false.

    $self->enable_test_mtime()
        Default true.

    $self->enable_test_ctime()
        Default true.

    $self->enable_test_permissions()
        Default false.

    $self->enable_test_special()
        Default true. If set false, it also sets `fifo' false.

    $self->enable_test_fifo()
        Default false. AKA named pipes.

    $self->enable_test_symlink()
        Default true, except for platforms that do not support symlinks (for
        example MSWin32 and cygwin) as determined by
        `$Config::Config{d_symlink}'.

    $self->enable_test_hardlink()
        AKA the `link()' function. Default true. If set false, this also
        sets `nlink' false.

    $self->enable_test_nlink()
        Count hard links. Default true.

    $self->enable_test_chown()
        Default false.

TEST CASES
    setup()
        Runs before every test to prepare a directory for testing.

    teardown()
        Runs after every test to clean up the test directory so the next
        test will have a clean workspace.

    Introduced($version)
        A subroutine attribute used to flag the Test::Virtual::Filesystem
        version number when that test was introduced. It's used like this:

          sub open_nonexistent_file : Tests(1) : Introduced('0.02') {
             ok(!open(my $f, '<', '/tmp/no_such_file'));
          }

    Features($featurelist)
        This is a subroutine attribute to specify one or more features used
        in the test. The features should be listed as a comma-separated
        list:

          sub symlink_create : Tests(1) : Features('symlink') {
             ok(symlink($src, $dest));
          }
          sub symlink_permissions : Tests(2) : Features('symlink, permissions') {
             ok(symlink($src, $dest));
             ok(-w $dest);
          }

        Subfeatures must be separated from their parent features by a `/'.
        For example:

          sub atime_mtime_set : Tests(1) : Features('time/atime, time/mtime') {
             my $now = time;
             ok(utime($now, $now, $file));
          }

        Look at the source code for `%feature_defaults' to see the supported
        features and subfeatures. The `enable_test_*' methods above describe
        the all the features, but in those methods the subfeature names are
        flattened.

    stat_dir(), introduced in v0.01
    read_dir(), introduced in v0.01
    read_dir_fail(), introduced in v0.01
    read_file_fail(), introduced in v0.01
    write_empty_file(), introduced in v0.01
    write_file(), introduced in v0.01
    write_file_subdir_fail(), introduced in v0.01
    write_append_file(), introduced in v0.01
    write_read_file(), introduced in v0.01
    write_read_file_binary(), introduced in v0.08
    write_unlink_file(), introduced in v0.01
    write_mkdir(), introduced in v0.01
    write_mkdir_fail(), introduced in v0.01
    write_rmdir(), introduced in v0.01
    write_subdir(), introduced in v0.01
    symlink_create(), introduced in v0.02
    symlink_follow(), introduced in v0.04
    symlink_deep(), introduced in v0.06
    symlink_loop(), introduced in v0.06
    truncate_file(), introduced in v0.06
    truncate_no_file(), introduced in v0.06
    truncate_file_no_dir(), introduced in v0.06
    truncate_dir(), introduced in v0.06
    time_mtime_create(), introduced in v0.06
    time_ctime_create(), introduced in v0.06
    time_mtime_set(), introduced in v0.06
    time_atime_set(), introduced in v0.06
    xattr_list(), introduced in v0.02
    xattr_set(), introduced in v0.02
    rename_file(), introduced in v0.08
    rename_file_exists(), introduced in v0.08
    rename_file_self(), introduced in v0.08
    rename_file_subdir(), introduced in v0.08
    rename_file_missing_src(), introduced in v0.08
    rename_file_missing_srcdir(), introduced in v0.08
    rename_file_missing_destdir(), introduced in v0.08
    rename_dir(), introduced in v0.08
    rename_dir_exists(), introduced in v0.08
    rename_dir_notempty(), introduced in v0.08
    rename_dir_self(), introduced in v0.08
    rename_dir_subdir(), introduced in v0.08
    rename_mismatch_dir(), introduced in v0.08
    rename_mismatch_file(), introduced in v0.08
    rename_symlink(), introduced in v0.08
CODE PHILOSOPHY
    These are some coding/design rules for the tests:

    Use only core filesystem functions
        Don't use File::Slurp, File::Path, etc. because they abstract
        filesystem operations and make it less clear what we're testing.

    Keep the tests small
        Test as little as possible in each method. Let authors know what's
        failed by the pattern of failing tests. This also helps avoid
        needing to edit the tests later.

    Avoid editing methods
        Don't break published CPAN code. If you want to test something new,
        write a new method.

    Try to use a few different filesystem functions as practical in one
    method
        For example, if you're testing `chmod', don't `mkdir' or `chown'
        unless you're writing a `chmod_mkdir_chown' test.

    Minimize test infrastructure
        Use method attributes and Test::Class features to keep the test
        methods really simple.

SEE ALSO
    Test::Class

    Fuse::PDF

AUTHOR
    Chris Dolan, *cdolan@cpan.org*



Changes
# History of Perl distribution Test-Virtual-Filesystem.  This change log is valid YAML.

0.12 :
     - Released: 2007 Dec 05
     - Summary: Fix for Cygwin

     - Fix: Cygwin, like MSWin32, allows rename(<dir>, <file>)
       Credit: cpan tester srezic@cpan.org
       CPANTesters: http://www.nntp.perl.org/group/perl.cpan.testers/834715

0.11 :
     - Released: 2007 Dec 04
     - Summary: Fix for Solaris

     - Feature: Supports Solaris

     - Fix: Solaris reports EEXIST, not ENOTEMPTY, for rename_dir_notempty
       Credit: cpan-tester Davd Cantrell
       CPANTesters: http://www.nntp.perl.org/group/perl.cpan.testers/2007/11/msg812452.html

0.10 :
     - Released: 2007 Dec 02
     - Summary: Fixes for Win32

     - Feature: Supports Win32 (well, at least Strawberry Perl 5.8.8)

     - Fix: MSWin32 allows rename(<dir>, <file>), instead of ENOTDIR
       Credit: cpan tester srezic@cpan.org
       CPANTesters: http://www.nntp.perl.org/group/perl.cpan.testers/2007/11/msg806443.html

     - Caveats: Fails rename_dir_notempty on Solaris.  Better diagnostics should make this fixable.
       Credit: cpan-tester Davd Cantrell
       CPANTesters: http://www.nntp.perl.org/group/perl.cpan.testers/2007/11/msg812452.html

0.09_01 :
     - Released: 2007 Nov 26
     - Summary: Speculative fixes for platform inconsistencies

     - Fix: MSWin32 reports EACCES, not EISDIR, for truncate_dir
       Credit: cpan tester srezic@cpan.org
       CPANTesters: http://www.nntp.perl.org/group/perl.cpan.testers/2007/11/msg806443.html

     - Fix: MSWin32 reports ENOENT, not ENOTDIR, for truncate_file_no_dir
       Credit: cpan tester srezic@cpan.org
       CPANTesters: http://www.nntp.perl.org/group/perl.cpan.testers/2007/11/msg806443.html

     - Fix: MSWin32 reports EACCES, not EISDIR, for rename_mismatch_dir
       Credit: cpan tester srezic@cpan.org
       CPANTesters: http://www.nntp.perl.org/group/perl.cpan.testers/2007/11/msg806443.html

     - Internals: Better failure diagnostics, including POSIX strerror() messages for $OS_ERROR

0.08 :
     - Released: 2007 Nov 26
     - Summary: Added rename tests

     - Feature: Added rename_* tests, documented some unimplemented corner cases

     - Prerequisites: Added Test::Builder::Tester v1.04
     - Prerequisites: Pointlessly add some core deps (Carp, File::Spec, POSIX) for CPANTS

     - Internals: Convert */*.t to use Test::Builder::Tester instead of ad-hoc Test::Builder data
     - Internals: Got rid of noisy messages about TODO and skip tests
     - Internals: Make read operations optionally die, as per test needs

0.07 :
     - Released: 2007 Nov 24
     - Summary: Fix nested symlink test for Linux; Forgive skew between clock and disk timestamp

     - Fix: Linux has a stricter limit of 5 nested symlinks.  Settle on this lower limit.
       Credit: cpan-testers David Cantrell and ICERIDER
       CPANTesters: http://www.nntp.perl.org/group/perl.cpan.testers/2007/11/msg791775.html      

     - Fix: Add tolerance for difference between CPU clock and disk timestamp
       Credit: cpan-tester David Cantrell
       CPANTesters: http://www.nntp.perl.org/group/perl.cpan.testers/2007/11/msg791775.html

     - Prerequisites: Added Readonly.pm v1.00

     - Press: http://www.nntp.perl.org/group/perl.qa/2007/11/msg9676.html

0.06 :
     - Released: 2007 Nov 20
     - Summary: Perl prereq to 5.8.0; new truncate, symlink, a/m/ctime tests

     - Incompatibility: Upped prereq to Perl 5.8.0 instead of 5.6.0
     - Incompatibility: (minor) renamed "enable_test_utime" to "enable_test_mtime"

     - Fix: Specify actual minimum of Perl 5.8.0; 5.6.x bombed on local through ref
       Credit: cpan-tester David Cantrell
       CPANTesters: http://www.nntp.perl.org/group/perl.cpan.testers/790313

     - Internals: Removed unneeded prereqs from Build.PL and Filesystem.pm

0.05 :
     - Released: 2007 Nov 19
     - Summary: Fix for non-English localized Perl

     - Fix: Check numeric $! against POSIX errno instead of string $! against english
       Credit: cpan-tester Andreas Koenig
       RT: http://rt.cpan.org/Ticket/Display.html?id=30839
       CPANTesters: http://www.nntp.perl.org/group/perl.cpan.testers/2007/11/msg784219.html

0.04 :
     - Released: 2007 Nov 18
     - Summary: Renamed test_* methods from v0.03; Subclassing; Bug fixes

     - Incompatibility: Renamed 'test_*' methods to 'enable_test_*'

     - Fix: Add List::MoreUtils to Build.PL/Makefile.PL
       Credit: cpan-tester David Cantrell
     - Fix: Turn off symlink and chown for Win32; better support for OS non-features
       Credit: cpan-tester stro@cpan.org, MSWin32 (for symlink part)
     - Fix: Enable subclassing (via a hack/workaround for Test::Class 0.24 -- see rt.cpan.org 30836)

     - Internals: Moved self-test tests into subclasses in t/*.t

     - TODO: Remove above workaround if RT 30836 is resolved

     - Press: http://use.perl.org/~ChrisDolan/journal/34920

0.03 :
     - Released: 2007 Nov 16
     - Summary: Important bug fixes

     - Fix: Don't use "\n" since it is one or two characters, depending on platform
       Credit: cpan-tester stro@cpan.org, MSWin32

     - Fix: Disable the stat_dir_size() test because it is too filesystem-dependent
       Credit: cpan-tester srezic@cpan.org, cygwin

     - Fix: Make xattr test more tolerant of getxattr() implementations (undef vs. '')

0.02 :
     - Released: 2007 Nov 16
     - Summary: Backward compatibility, xattrs, symlinks

     - Feature: let client specify a compatibility version.  Mark newer tests TODO.
     - Feature: skip tests that the client wants to ignore
     - Feature: new test to exercise symlinks
     - Feature: new test to read/write extended attributes
     - Feature: new test to check that dirs have non-zero size

     - Prerequisites: (optional) File::ExtAttr v0.08

     - Press: http://use.perl.org/article.pl?sid=07/11/16/0925222

0.01 :
     - Released: 2007 Nov 14
     - Summary: initial release

Build.PL
use Module::Build;

Module::Build->new(
   module_name => 'Test::Virtual::Filesystem',
   dist_author => 'Chris Dolan <cdolan@cpan.org>',
   license => 'perl',
   requires => {
      'perl' => '5.8.0',
      'List::MoreUtils' => 0,
      'Test::Class' => '0.24',
      'Test::More' => 0.62,
      'Attribute::Handlers' => '0.78', # implied by Test::Class, but just in case...
      'Readonly' => 1.00,
      'File::Spec' => 0,
      'POSIX' => 0,
      'Carp' => 0,
   },
   build_requires => {
      'Test::Builder::Tester' => '1.04',
   },
   recommends => {
      # Optional for testing extended attributes
      'File::ExtAttr' => 1,

      # Optional in tests, skipped if not present
      'Test::Pod' => 0,
      'Test::Pod::Coverage' => 0,
   },
   create_readme => 1,
   create_makefile_pl => 'traditional',
)->create_build_script;


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.03
use ExtUtils::MakeMaker;
WriteMakefile
(
          'NAME' => 'Test::Virtual::Filesystem',
          'VERSION_FROM' => 'lib/Test/Virtual/Filesystem.pm',
          'PREREQ_PM' => {
                           'Attribute::Handlers' => '0.78',
                           'Carp' => '0',
                           'File::Spec' => '0',
                           'List::MoreUtils' => '0',
                           'POSIX' => '0',
                           'Readonly' => '1',
                           'Test::Builder::Tester' => '1.04',
                           'Test::Class' => '0.24',
                           'Test::More' => '0.62'
                         },
          'INSTALLDIRS' => 'site',
          'EXE_FILES' => [],
          'PL_FILES' => {}
        )
;
